lib/src/Pos/Worker/Block.hs

```diff
@@ -146,27 +149,139 @@ blockCreator
+                                   (blockHeaderHash tipHeader)
+            case mbEbbBlund of
+                Nothing -> Prelude.error "Pos.Worker.Block.dropObftEbb: unable to get blund for EBB"
+                Just ebbBlund -> do
+                    rollbackBlocks genesisConfig $ NewestFirst (ebbBlund :| [])
+                    normalizeMempool genesisConfig txpConfig
+
+blockCreatorOriginal
+    :: ( BlockWorkMode ctx m
+       , HasMisbehaviorMetrics ctx
+       )
+    => Genesis.Config
+    -> TxpConfiguration
+    -> SlotId
+    -> Diffusion m -> m ()
+blockCreatorOriginal genesisConfig txpConfig (slotId@SlotId {..}) diffusion = do
     -- First of all we create genesis block if necessary.
     mGenBlock <- createGenesisBlockAndApply genesisConfig txpConfig siEpoch
-    whenJust mGenBlock $ \createdBlk -> do
-        logInfo $ sformat ("Created genesis block:\n" %build) createdBlk
-        jsonLog $ jlCreatedBlock (configEpochSlots genesisConfig) (Left createdBlk)
-
-    -- Then we get leaders for current epoch.
-    leadersMaybe <- LrcDB.getLeadersForEpoch siEpoch
-    case leadersMaybe of
-        -- If we don't know leaders, we can't do anything.
-        Nothing -> logWarning "Leaders are not known for new slot"
-        -- If we know leaders, we check whether we are leader and
-        -- create a new block if we are. We also create block if we
-        -- have suitable PSK.
-        Just leaders ->
-            maybe onNoLeader
-                  (onKnownLeader leaders)
-                  (leaders ^? ix (fromIntegral $ getSlotIndex siSlot))
+
+    -- the ConsensusEra could've changed due to to this
+    -- call trace (thing above calls thing below):
+
+    --    blockCreatorOriginal
+    -- -> createGenesisBlockAndApply
+    -- -> createGenesisBlockDo
+    -- -> verifyBlocksPrefix
+    -- -> usVerifyBlocks
+    -- -> verifyBlock
+    -- -> processGenesisBlock
+    -- -> adoptBlockVersion
+
+    era <- getConsensusEra
+    case era of
+        OBFT _ -> blockCreator genesisConfig txpConfig slotId diffusion
+        Original -> do
+            whenJust mGenBlock $ \createdBlk -> do
+                logInfo $ sformat ("Created genesis block:\n" %build) createdBlk
+                jsonLog $ jlCreatedBlock (configEpochSlots genesisConfig) (Left createdBlk)
+
+            -- Then we get leaders for current epoch.
+            leadersMaybe <- LrcDB.getLeadersForEpoch siEpoch
+            case leadersMaybe of
+                -- If we don't know leaders, we can't do anything.
+                Nothing -> logWarning "Leaders are not known for new slot"
+                -- If we know leaders, we check whether we are leader and
+                -- create a new block if we are. We also create block if we
+                -- have suitable PSK.
+                Just leaders ->
+                    maybe onNoLeader
+                          (onKnownLeader leaders)
+                          (leaders ^? ix (fromIntegral $ getSlotIndex siSlot))
```

Comment:
Context for other reviewers: this part is just the red deletions above, indented over. We hide the [`whenJust`](https://github.com/input-output-hk/cardano-sl/pull/4077/files#diff-e80ac54ce327ad386ed3ba9fe2fa4313R269) in this branch of the `case`, so that creation of the EBB won't be reported (since we will roll it back shortly).